draft/ZEP0010.md

```diff
@@ -0,0 +1,440 @@
+Where possible, however, implementations are encouraged, to delegate
+that logic via a callback or plugin mechanism that allows third-party code to
+handle the extension dynamically.
+
+As the set of extensions evolves, certain interfaces may arise which allow
+this modular approach for a subset of extensions. Where possible, these
+interfaces will be added to the specification. Feedback from implementers
+on such matters is highly encouraged.
+
+### Examples
+
+The following examples represent a few realistic use cases of the top-level
+``extensions`` container. This ZEP is putting in place the mechanism so the
+community can experiment with such extensions before their standardization.
+
+#### Offset (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.offset",
+            "configuration": { "offset": [ 12, 24 ] }
+        }
+    ]
+}
+```
+
+The ``example.offset`` extension contains an array of the same order as the
+shape of the containing array specifying which element of the array should be
+considered as the origin, e.g., `[0, 0]`. This allows the reuse of subregions
+of an array without the need to rewrite the data.
+
+Note that in this example of the extension is ``must_understand=true`` meaning
+an implementation which does not support the ``example.offset`` extension
+should raise an error.
+
+#### Statistics (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.array-statistics",
+            "must_understand": false,
+            "configuration": {
+                "min": 5,
+                "max": 1023
+            }
+        }
+    ]
+}
+```
+
+The ``example.array-statistics`` extension contains two fields -- ``min``
+and ``max`` specifying the range of values which are present in the array,
+reducing the need to read every byte. ``must_understand`` is false, so
+implementations can safely ignore the extension.
+
+#### Skip empty chunks (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    ...,
+    "extensions": [
+        "example.skip_empty_chunks"
+    ]
+}
+```
+
+Currently the "write_empty_chunks" flag in zarr-python is not propagated
+to the zarr.json file. An extension like ``example.skip_empty_chunks``
+could serve as a no-configuration flag in the metadata to inform
+implementations that empty chunks should not be written.
+
+
+
+#### Multiscale arrays (group)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "group",
+    ...,
+    "extensions": [
+        {
+            "name": "example.multiscale-arrays",
+            "must_understand": false,
+            "configuration": {
+                "multiscale": {
+                    "datasets": [
+                        "path/to/array/1",
+                        "path/to/array/2",
+                        "path/to/array/3"
+                    ]
```

Comment:
Could you provide some details about why using extensions is better than using attributes for this purpose? I'm asking this because it seems to serve a similar purpose as the [NGFF spec](https://ngff.openmicroscopy.org/0.5/index.html) or proposed GeoZarr spec which have been building an attributes-centric approach.

Reply:
as an example of the motivation for this question and the Zarr v3 restriction, I'm trying to understand how this could be used for GeoZarr (if the SWG decides to use this mechanism) which needs to support both Zarr formats. An example following my current understanding is drafted at https://github.com/maxrjones/geozarr-spec/blob/simple-translation/geozarr-spec.md#defining-primary-geospatial-convention.

Reply:
> This example is trying to outline how those two communities might move their common data structures into a shared space. GeoZarr's similar structure was initially taken from https://github.com/zarr-developers/zarr-specs/issues/50. Had we had the central registry at that time, we might have developed it in common.

Does this mean that ZEP009 supercedes [ZEP004](https://zarr.dev/zeps/draft/ZEP0004.html)?

> the value of a central registry seems like a separate question from why using extensions is better than using attributes.

My guess based on these discussion is that an extensions key allows the Zarr community to watch for naming conflicts between communities, serving a similar purpose as ZEP0004 but with a different mechanism. Is this correct?

Reply:
> As the person who originated ZEP 4, I have come to feel it wasn't fully baked. I feel like the mechanism for identifying if a convention is present proposed there is quite flaky. I much prefer the STAC approach, i.e. explicit extensions, which is closer to what is being described here.

What I had in mind was explicitly registering individual top-level attribute names, e.g. `tensorstore.dimension_units`.